File: code/backend-php/src/Adapter/Auth/Rest/GuestAuthController.php
<?php

namespace Games\Paths\Adapter\Auth\Rest;

use Games\Paths\Core\Port\Auth\GuestAuthPort;
use Games\Paths\Core\Port\Auth\JwtPort;
use Games\Paths\Core\Port\Auth\TokenPersistencePort;
use Games\Paths\Core\Domain\Auth\GuestSession;
use Games\Paths\Adapter\Rest\Cookie\CookieHelper;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Controller for guest authentication (Step 12/13).
 */
class GuestAuthController
{
    private GuestAuthPort $guestAuthService;
    private JwtPort $jwtPort;
    private TokenPersistencePort $tokenPersistence;

    public function __construct(
        GuestAuthPort $guestAuthService,
        JwtPort $jwtPort,
        TokenPersistencePort $tokenPersistence
    ) {
        $this->guestAuthService = $guestAuthService;
        $this->jwtPort = $jwtPort;
        $this->tokenPersistence = $tokenPersistence;
    }

    /**
     * Build the session response — Step 13 version (tokens in cookies).
     */
    private function processSessionResponse(Response $response, GuestSession $session): array
    {
        $uuid = $session->getUuid();
        $username = $session->getUsername();
        $role = $session->getRole();

        // 1. Generate real JWT tokens
        $accessToken = $this->jwtPort->generateAccessToken($uuid, $username, $role);
        $refreshToken = $this->jwtPort->generateRefreshToken($uuid);

        // 2. Persist refresh token in DB (Step 13)
        $tokenInfo = $this->jwtPort->parseToken($refreshToken);
        $this->tokenPersistence->saveRefreshToken($uuid, $refreshToken, $tokenInfo);

        // 3. Set HttpOnly cookies
        // Refresh Token
        $response = CookieHelper::setCookie(
            $response,
            'pathsgames.refreshToken',
            $refreshToken,
            7 * 24 * 60 * 60, // 7 days
            '/'
        );

        // Guest Cookie Token (for resume)
        $response = CookieHelper::setCookie(
            $response,
            'pathsgames.guestcookie',
            $session->getCookieToken(),
            30 * 24 * 60 * 60, // 30 days
            '/'
        );

        // 4. Return clean JSON (no refresh token, no guest cookie token)
        return [
            'response' => $response,
            'data' => [
                'userUuid' => $uuid,
                'username' => $username,
                'role'     => $role,
                'accessToken' => $accessToken,
                'accessTokenExpiresAt' => $this->jwtPort->getAccessTokenExpirationMs(),
                'refreshTokenExpiresAt' => $this->jwtPort->getRefreshTokenExpirationMs(),
            ]
        ];
    }

    public function createGuest(Request $request, Response $response): Response
    {
        $session = $this->guestAuthService->createGuestSession();
        $result = $this->processSessionResponse($response, $session);
        
        $newResponse = $result['response'];
        $newResponse->getBody()->write(json_encode($result['data']));
        
        return $newResponse->withHeader('Content-Type', 'application/json')->withStatus(201);
    }

    public function resumeGuest(Request $request, Response $response): Response
    {
        // Step 13: Read guestCookieToken from HttpOnly cookie if body is empty
        $cookies = $request->getCookieParams();
        $guestCookieToken = $cookies['pathsgames.guestcookie'] ?? null;

        if (!$guestCookieToken) {
            // Fallback to body for backward compatibility or explicit requests
            $data = json_decode((string)$request->getBody(), true);
            $guestCookieToken = $data['guestCookieToken'] ?? null;
        }

        if (!$guestCookieToken) {
            $error = ['error' => 'Bad Request', 'message' => 'Missing guestCookieToken (cookie or body)'];
            $response->getBody()->write(json_encode($error));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
        }

        $session = $this->guestAuthService->resumeGuestSession($guestCookieToken);
        if (!$session) {
            $error = [
                'error' => 'SESSION_EXPIRED_OR_NOT_FOUND',
                'message' => 'Guest session expired or not found. Please create a new session.'
            ];
            $response->getBody()->write(json_encode($error));
            return $response->withHeader('Content-Type', 'application/json')->withStatus(401);
        }

        $result = $this->processSessionResponse($response, $session);
        
        $newResponse = $result['response'];
        $newResponse->getBody()->write(json_encode($result['data']));
        
        return $newResponse->withHeader('Content-Type', 'application/json')->withStatus(200);
    }
}

File: code/backend/python/app/adapters/rest/auth/guest_auth_controller.py
from fastapi import APIRouter, HTTPException, status, Request, Response, Header
from pydantic import BaseModel
from typing import Optional
from app.core.ports.auth.guest_auth_port import GuestAuthPort
from app.core.ports.auth.jwt_port import JwtPort
from app.core.ports.auth.token_persistence_port import TokenPersistencePort

from fastapi.responses import JSONResponse
from app.core.services.security import rate_limit_service as _rl

RATE_BUCKET = "guest"

class GuestResumeRequest(BaseModel):
    guestCookieToken: Optional[str] = None

class GuestAuthController:
    def __init__(self, guest_auth_port: GuestAuthPort, jwt_port: JwtPort, token_persistence: TokenPersistencePort,
                 test_endpoints_enabled: bool = False, rate_limit_service=None, guest_per_ip: int = 0,
                 csrf_token_service=None):
        self.guest_auth_port = guest_auth_port
        self.jwt_port = jwt_port
        self.token_persistence = token_persistence
        self.test_endpoints_enabled = test_endpoints_enabled
        # v0.37.7 — Step 41: the guest bucket of the rate limiter and the CSRF token issuer
        self.rate_limit_service = rate_limit_service
        self.guest_per_ip = guest_per_ip
        self.csrf_token_service = csrf_token_service
        self.router = APIRouter(prefix="/api/auth/guest")
        self.router.add_api_route("", self.create_guest, methods=["POST"], status_code=status.HTTP_201_CREATED)
        self.router.add_api_route("/resume", self.resume_guest, methods=["POST"])

    def _process_session_response(self, session):
        uuid = session.user_uuid
        username = session.username
        role = "PLAYER" # Guest default

        # 1. Generate JWT tokens
        access_token = self.jwt_port.generate_access_token(uuid, username, role)
        refresh_token = self.jwt_port.generate_refresh_token(uuid)

        # 2. Persist refresh token (Step 13)
        token_info = self.jwt_port.parse_token(refresh_token)
        self.token_persistence.save_refresh_token(uuid, refresh_token, token_info)

        # 3. Create JSON Response
        data = {
            "userUuid": uuid,
            "username": username,
            "role": role,
            "accessToken": access_token,
            "accessTokenExpiresAt": self.jwt_port.get_access_token_expiration_ms(),
            "refreshTokenExpiresAt": self.jwt_port.get_refresh_token_expiration_ms()
        }
        if self.csrf_token_service is not None:
            data["csrfToken"] = self.csrf_token_service.token_for(access_token)
        
        response = JSONResponse(content=data)

        # 4. Set HttpOnly cookies with SameSite=None and Secure=True
        response.set_cookie(
            key="pathsgames.refreshToken",
            value=refresh_token,
            httponly=True,
            samesite="none",
            secure=True,
            path="/",
            max_age=180 * 24 * 60 * 60  # 6 months
        )

        response.set_cookie(
            key="pathsgames.guestcookie",
            value=session.guest_cookie_token,
            httponly=True,
            samesite="none",
            secure=True,
            path="/",
            max_age=180 * 24 * 60 * 60  # 6 months
        )

        return response

    def create_guest(self, request: Request,
                     x_test_marker: Optional[str] = Header(default=None)):
        # The X-Test-Marker header tags the guest as test data so it can be
        # removed by POST /api/dev/cleanup. Honoured only when dev test
        # endpoints are enabled; ignored in production.
        # Step 41 — at most guest_per_ip new guests per source address and window
        if self.rate_limit_service is not None and self.guest_per_ip > 0:
            ip = _rl.client_ip(request.headers.get("X-Forwarded-For"),
                               request.client.host if request.client else None)
            verdict = self.rate_limit_service.try_acquire(RATE_BUCKET, ip, self.guest_per_ip)
            if not verdict.allowed:
                return _rate_limited(verdict, "Too many guest sessions from this address")
        marker = x_test_marker if self.test_endpoints_enabled else None
        session = self.guest_auth_port.create_guest_session(marker)
        response = self._process_session_response(session)
        response.status_code = status.HTTP_201_CREATED
        return response

    def resume_guest(self, request_in: Request, body: Optional[GuestResumeRequest] = None):
        # Step 13: Read from cookie first
        guest_cookie_token = request_in.cookies.get("pathsgames.guestcookie")
        
        # Fallback to body
        if not guest_cookie_token and body:
            guest_cookie_token = body.guestCookieToken

        if not guest_cookie_token:
            raise HTTPException(status_code=400, detail="Missing guestCookieToken (cookie or body)")

        session = self.guest_auth_port.resume_guest_session(guest_cookie_token)
        if not session:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail={
                    "error": "SESSION_EXPIRED_OR_NOT_FOUND",
                    "message": "Guest session expired or not found. Please create a new session."
                }
            )
        return self._process_session_response(session)


def _rate_limited(verdict, what: str) -> JSONResponse:
    import time
    return JSONResponse(
        status_code=429,
        headers={"Retry-After": str(verdict.retry_after_seconds)},
        content={"error": "RATE_LIMITED",
                 "message": f"{what}, retry in {verdict.retry_after_seconds} seconds",
                 "retryAfterSeconds": verdict.retry_after_seconds,
                 "timestamp": int(time.time() * 1000)},
    )
